elobot.py

```diff
@@ -54,245 +29,259 @@ def get_channel_id(self, channel_name):
             if channel['name'] == channel_name:
                 return channel['id']
 
-        print('Unable to find channel: ' + channel_name)
-        quit()
+        raise ValueError(f'No channel "{channel_name}"')
 
-class EloBot(object):
-    players = defaultdict(Player)
 
+class EloBot:
     def __init__(self, slack_client, channel_id, name, min_streak_len):
         self.name = name
         self.slack_client = slack_client
         self.min_streak_len = min_streak_len
         self.channel_id = channel_id
 
-        self.last_ping = 0
-
-        self.init_players()
-        self.ensure_connected()
-        self.run()
-
-    def rank_game(self, winner, loser):
-        # From https://metinmediamath.wordpress.com/2013/11/27/how-to-calculate-the-elo-rating-including-example/
-        winner_transformed_rating = 10 ** (winner.rating / 400.0)
-        loser_transformed_rating  = 10 ** (loser.rating  / 400.0)
-
-        winner_expected_score = winner_transformed_rating / (winner_transformed_rating + loser_transformed_rating)
-        loser_expected_score  = loser_transformed_rating  / (winner_transformed_rating + loser_transformed_rating)
-
-        winner_new_elo = round(winner.rating + winner.k_factor * (1 - winner_expected_score))
-        loser_new_elo  = round(loser.rating  + loser.k_factor  * (0 - loser_expected_score))
-
-        winner_elo_delta = winner_new_elo - winner.rating
-        loser_elo_delta  = loser_new_elo  - loser.rating
-
-        winner.wins += 1
-        loser.losses += 1
-
-        winner.rating = winner_new_elo
-        loser.rating = loser_new_elo
-
-        return winner_elo_delta, loser_elo_delta
-
-    def apply_match(self, match):
-        """
-        Apply a match to the ranking system, changing winner and loser's elos.
-        Return (winner elo delta, loser elo delta)
-        """
-        if not match.pending:
-            raise ValuError("Match must be pending to apply.")
-
-        with db.transaction():
-            winner = self.players[match.winner_handle]
-            loser  = self.players[match.loser_handle]
-            winner_elo_delta, loser_elo_delta = self.rank_game(winner, loser)
-
-            match.pending = False
-            match.save()
+        # TODO: Should map from Player to int. Handles should be dealt with as little as possible
+        # Map from handle to elo_cache
+        self.elo_cache = defaultdict(lambda: 1500)  # Default ELO is 1500
```

Comment:
by introducing this you are once again introducing a second source of truth for ELOs

Reply:
...How? All the elo cache does is keep track of the elo deltas since the last time they've been displayed to the Slack channel.

Perhaps I don't quite understand what you mean by "source of truth"?